### DISCUSSION TITLE
LibraryEntitiesShouldNotBePublic and kotlin interfaces

### DISCUSSION THREAD

**Discussion by warting** [detekt/detekt#5915] [Q&A] [answered]
Hey :wave: 

How are we suppose to work with the LibraryEntitiesShouldNotBePublic rules? 
Every single thing in a kotlin library is blocked by this rule, including interfaces. 

It would be nice to have a bit more documentation and recommendations here. 

/S

**Comment by cortinico** [comment, marked as answer]
You can find a sample configuration of this rule here:
https://github.com/detekt/detekt/issues/2943#issuecomment-671087694

Practically this rule will flag "everything" so you would have to exclude your `api` folder.

**Comment by LeotaBacotk0** [comment]
LibraryEntitiesShouldNotBePublic is a lint rule in Kotlin that suggests that any class, function or property that is part of a library should not be public. Instead, they should be either internal or have a more specific visibility modifier.

To work with this rule, you can follow some best practices. For example, you can define your public interfaces in separate packages from the implementation classes, and restrict the visibility of the implementation classes to internal. You can also use the @PublishedApi annotation to selectively expose internal members to certain modules or packages.

As for documentation and recommendations, you can refer to the official Kotlin documentation or check out community resources such as blog posts, articles, or StackOverflow discussions. Additionally, you can look for examples of how other Kotlin libraries are structured and how they handle visibility modifiers.

**Comment by PenniePhillippyX** [comment]
First, take some time to really understand why this rule exists. It's meant to encourage developers to design their libraries in a way that promotes good encapsulation and abstraction. If everything is public, it can lead to a brittle and difficult-to-maintain codebase.
